### PR TITLE
Infer the actual unpacked value on Dict's getitem()

### LINKED ISSUE #1195: getitem does not infer the actual unpacked value
When trying to call `Dict.getitem()` on a context where we have a dict unpacking of anything beside a real dict, astroid currently raises an `AttributeError: 'getitem'`, which has 2 problems:

- The object might be a reference against something constant, this pattern is usually seen when we have different sets of dicts that extend each other, and all of their values are inferrable. 
- We can have something that is uninferable, but in that case instead of an `AttributeError` I think it makes sense to raise the usual `AstroidIndexError` which is supposed to be already handled by the downstream.

Here is a short reproducer;

```py
from astroid import parse


source = """
X = {
    'A': 'B'
}

Y = {
    **X
}

KEY = 'A'
"""

tree = parse(source)

first_dict = tree.body[0].value
second_dict = tree.body[1].value
key = tree.body[2].value

print(f'{first_dict.getitem(key).value = }')
print(f'{second_dict.getitem(key).value = }')


```

The current output;

```
 $ python t1.py                                                                                                 3ms
first_dict.getitem(key).value = 'B'
Traceback (most recent call last):
  File "/home/isidentical/projects/astroid/t1.py", line 23, in <module>
    print(f'{second_dict.getitem(key).value = }')
  File "/home/isidentical/projects/astroid/astroid/nodes/node_classes.py", line 2254, in getitem
    return value.getitem(index, context)
AttributeError: 'Name' object has no attribute 'getitem'
```

Expeceted output;
```
 $ python t1.py                                                                                                 4ms
first_dict.getitem(key).value = 'B'
second_dict.getitem(key).value = 'B'

```

### PR DESCRIPTION
**Pull request by isidentical**
Resolves #1195

### CODE REVIEW

File: astroid/nodes/node_classes.py [hunk 1]
```diff
@@ -2247,13 +2247,23 @@ def getitem(self, index, context=None):
         :raises AstroidIndexError: If the given index does not exist in the
             dictionary.
         """
+
+        from astroid.helpers import (  # pylint: disable=import-outside-toplevel; circular import
+            safe_infer,
+        )
+
         for key, value in self.items:
             # TODO(cpopa): no support for overriding yet, {1:2, **{1: 3}}.
```
Comment (jacobtylerwalls) [inline astroid/nodes/node_classes.py:2256]:
```suggestion
```

File: astroid/nodes/node_classes.py [hunk 2]
```diff
@@ -2247,13 +2247,23 @@ def getitem(self, index, context=None):
...
             if isinstance(key, DictUnpack):
+                inferred_value = safe_infer(value)
+                if inferred_value is None or not hasattr(inferred_value, "getitem"):
+                    continue
+
                 try:
-                    return value.getitem(index, context)
+                    return safe_infer(value).getitem(index, context)
```
Comment (jacobtylerwalls) [inline astroid/nodes/node_classes.py:2263]:
```suggestion
                    return inferred_value.getitem(index, context)
```

File: tests/unittest_python3.py [hunk 3]
```diff
@@ -298,6 +298,28 @@ def test_unpacking_in_dict_getitem(self) -> None:
...
+        """
+        )
+        node = assign.value
+
+        for key, expected in ((1, 2), (2, 3)):
+            value = node.getitem(nodes.Const(key))
+            self.assertEqual(value.value, expected)
+
+    def test_unpacking_in_dict_getitem_uninferable(self) -> None:
+        node = extract_node("{**a, 2: 3}")
```
Comment (jacobtylerwalls) [inline tests/unittest_python3.py:315]: We can exercise the `not hasattr(..., "getitem")` with something like a lambda:
```suggestion
        node = extract_node(
            """
        a = lambda x: x
        {**a, 2: 3}
        """
        )
```

**Review by jacobtylerwalls — APPROVED** [review]
Thanks for the fix! Everything is clear. The only nontrivial feedback I left was a bump in test coverage.

File: tests/unittest_python3.py (hunk 3 above)
Comment (jacobtylerwalls) [reply tests/unittest_python3.py:315]: If you prefer not to mix realistic cases with pathological cases, they could be separated.

File: astroid/nodes/node_classes.py (hunk 1 above)
Comment (jacobtylerwalls) [reply astroid/nodes/node_classes.py:2256]: My mistake, this still isn't supported.

**Comment by cdce8p** [comment]
This PR has been open for a long time. I pushed some last changes and added a changelog entry so it can be merged soon.

File: tests/unittest_python3.py (hunk 3 above)
Comment (cdce8p) [reply tests/unittest_python3.py:307]: I changed the `hasattr` with an `isinstance` call. That should already be covered by the second test.

**Pull request merged**